// @oasislabs/gateway/dist/lib/src/api.js
/**
 * The api module defines the types used for the v0/api of the developer
 * gateway.
 */
/**
 * Deploys a services.
 */
export var DeployApi = {
    url: 'v0/api/service/deploy',
    method: 'POST',
};
/**
 * Invokes an rpc on a service.
 */
export var RpcApi = {
    url: 'v0/api/service/execute',
    method: 'POST',
};
/**
 * Retrieives the public key for a service.
 */
export var PublicKeyApi = {
    url: 'v0/api/service/getPublicKey',
    method: 'POST',
};
/**
 * Retrieves the code for an address.
 */
export var GetCodeApi = {
    url: 'v0/api/service/getCode',
    method: 'POST',
};
/**
 * Polls for an event on a service.
 */
export var ServicePollApi = {
    url: 'v0/api/service/poll',
    method: 'POST',
};
/**
 * Subscribes to a gateway topic.
 */
export var SubscribeApi = {
    url: 'v0/api/event/subscribe',
    method: 'POST',
};
/**
 * Polls for an event on a subscription.
 */
export var SubscribePollApi = {
    url: 'v0/api/event/poll',
    method: 'POST',
};
/**
 * Retrieves the health status of the gateway.
 */
export var HealthApi = {
    url: 'v0/api/health',
    method: 'GET',
};
/**
 * Unsubscribes from a `SubscribeApi` subscription.
 */
export var UnsubscribeApi = {
    url: 'v0/api/event/unsubscribe',
    method: 'POST',
};
